internal/transports/p2p/server.go

```diff
@@ -29,99 +43,271 @@ func NewServer(
 	log *zerolog.Logger,
 ) *server {
 	serverLogger := log.With().Str("service", "p2p-experimental").Logger()
+	ctx, ctxCancel := context.WithCancel(context.Background())
 	server := &server{
 		config:         config,
 		chainParams:    chainParams,
 		headersService: headersService,
 		chainService:   chainService,
 		log:            &serverLogger,
-		peers:          make([]*peer.Peer, 0),
+		outboundPeers:  peer.NewPeersCollection(config.MaxOutboundConnections),
+		inboundPeers:   peer.NewPeersCollection(config.MaxInboundConnections),
+		addresses:      network.NewAdressBook(config.BanDuration, config.AcceptLocalPeers),
+		ctx:            ctx,
+		ctxCancel:      ctxCancel,
 	}
+
 	return server
 }
 
+// Start starts the P2P server by connecting to outbound peers and listening for inbound connections.
 func (s *server) Start() error {
-	err := s.seedAndConnect()
+	err := s.connectOutboundPeers()
 	if err != nil {
 		return err
 	}
 
-	return s.listenAndConnect()
+	err = s.listenInboundPeers()
+	if err != nil {
+		s.log.Error().Msgf("error during server start. Shutdown p2p server. Reason: %v", err)
+		s.Shutdown()
+
+		return err
+	}
+	return nil
 }
 
-func (s *server) Shutdown() error {
-	for _, p := range s.peers {
+// Shutdown gracefully shuts down the P2P server by disconnecting all peers.
+func (s *server) Shutdown() {
+	// Stop listening for incoming connections
+	if s.listener != nil {
+		_ = s.listener.Close()
+	}
+
+	// Cancel all child goroutines
+	s.ctxCancel()
+	s.ctxWg.Wait()
+
+	// Disconnect active peers
+	for _, p := range s.outboundPeers.Enumerate() {
+		p.Disconnect()
+	}
+	for _, p := range s.inboundPeers.Enumerate() {
 		p.Disconnect()
 	}
-	return nil
 }
 
-func (s *server) seedAndConnect() error {
+func (s *server) connectOutboundPeers() error {
 	seeds := seedFromDNS(s.chainParams.DNSSeeds, s.log)
 	if len(seeds) == 0 {
 		return errors.New("no seeds found")
 	}
 
+	if len(seeds) > int(s.config.MaxOutboundConnections) {
+		seeds = seeds[:s.config.MaxOutboundConnections]
+	}
```

Comment:
I think we should add the rest of the addresses from seeds to an AddressBook

Reply:
Also, I'm wondering if we should also filter the seed addresses with our `IsRoutable` function from `network` 🤔 . But maybe we should just trust them, IDK

Reply:
Good point. Now they are stored in address book and filtered with IsRoutable by the book